### PR TITLE
Don't use a relative path to resolve Default404 template

### PR DESCRIPTION
**Pull request by jviotti**
The current approach assumes that the current directory when running
`react-static` is the root path, which might not be the case if you
customize the root.

This came up as I'm developing a CLI module that embeds `react-static`,
takes some command line arguments, spawns `react-static` using files
inside the CLI module, and stores the results in the user's destination
directory.

When the tool (and therefore `react-static`) are installed globally, my
`ROOT` is:

```
/Users/jviotti/.nvm/versions/node/v9.11.2/lib/node_modules/<my-cli>
```

And `__dirname` from `src/static/getRoutes.js` becomes:

```
/Users/jviotti/.nvm/versions/node/v9.11.2/lib/node_modules/<my-cli>/node_modules/react-static/lib/static
```

Given those two variables, the code in `getRoutes.js` will evaluate the
`Default404` template path to:

```
node_modules/react-static/lib/browser/components/Default404
```

Which is a relative path, therefore `react-static` will try to find
`Default404` in a local installation of `react-static` in the place the
user ran the CLI, which is not expected.

Making the path absolute ensures that `react-static` always defaults the
404 template to the one it itself ships, instead of accidentally cross
referencing another version of itself.

Signed-off-by: Juan Cruz Viotti <jv@jviotti.com>

### CODE REVIEW

**Comment by sw-yx** [comment]
hi everyone! thanks for your PR! there is now a call for maintainers of React Static, please weigh in here: https://github.com/nozzle/react-static/issues/1259

**Comment by SleeplessByte** [comment]
@jviotti can you merge-in master. I think #1219 might have fixed Travis!

**Comment by jviotti** [comment]
@SleeplessByte Done, but I don't think Travis is kicking in anymore

File: packages/react-static/src/static/getRoutes.js [hunk 1]
```diff
@@ -46,12 +46,9 @@ export default async function getRoutes(state, callback = d => d) {
     if (!has404 && !incremental) {
       allNormalizedRoutes.unshift({
         path: '404',
-        template: path.relative(
-          state.config.paths.ROOT,
```
Comment (SleeplessByte) [inline packages/react-static/src/static/getRoutes.js]: Why was this removed?

`resolve` should probably _still_ use this `root` path?

(p.s. should now be running travis again).

Comment (jviotti) [reply packages/react-static/src/static/getRoutes.js]: The result of `path.resolve(__dirname, path.join('..', 'browser', 'components', 'Default404'))` is an absolute path, which always works fine when passed to `template`. 

Adding `path.relative(ROOT, <path>)` will calculate `<path>` relative to `ROOT`, so the path will be the same, just that now it will have a ton of `../../../` to point to the same directory from the point of view of `ROOT`. 

In the end, the result should be the same, as long as your current working directory is the same as `ROOT`. After this change, I believe everything that worked in the past should still work (lets see what Travis thinks), and it will now still work even if `process.cwd() !== ROOT`:

```sh
$ node
> path.relative('/foo/bar/baz', '/foo/bar/baz/qux')
'qux'
> path.relative('/foo/bar/baz', '/test/bar/baz/qux')
'../../../test/bar/baz/qux'
```

**Comment by jviotti** [comment]
@SleeplessByte Not sure if its just me, but I seem to be getting the same errors Travis is reporting here on `master`, which seems to explain why they look unrelated to my changes here. For example:

```
    TypeError: Cannot read property 'endsWith' of undefined

      142 |
      143 |   // add trailing slash only if assetsPath was supplied, but no trailing slash
    > 144 |   if (assetsPath !== '' && !assetsPath.endsWith('/')) {
          |                                        ^
      145 |     assetsPath = `${assetsPath}/`
      146 |   }
      147 |

      at endsWith (src/static/getConfig.js:144:40)
      at Object.<anonymous> (src/static/__tests__/getConfig.test.js:62:21)
```

Can you confirm it happens on master as well?

**Comment by jviotti** [comment]
@SleeplessByte Looks so. Consider this one:

```
    Snapshot name: `BodyWithMeta when route is a static route 1`

    - Snapshot
    + Received

    @@ -10,11 +10,11 @@
          </div>
          <script
            dangerouslySetInnerHTML={
              Object {
                "__html": "
    -     window.__routeInfo = {"routeDate":"here"};",
    +     window.__routeInfo = JSON.parse("{\\"routeDate\\":\\"here\\"}");",
              }
            }
            type="text/javascript"
          />
          <script
```

The test at `packages/react-static/src/static/components/__tests__/__snapshots__/BodyWithMeta.test.js.snap` definitely expects

```
window.__routeInfo = {\\"routeDate\\":\\"here\\"};",
```

While `packages/react-static/src/static/components/BodyWithMeta.js` has:

```
window.__routeInfo = JSON.parse(${jsesc(JSON.stringify(embeddedRouteInfo), {
```

File: packages/react-static/src/static/getRoutes.js (hunk 1 above)
Comment (SleeplessByte) [reply packages/react-static/src/static/getRoutes.js]: Thanks for the explanation! 

I didn't really mean, why was `path.relative` removed, but why was `state.config.paths.ROOT` removed. I think we used to be able to change the `ROOT` dir. Overall, I don't really care. All the files _should_ live under `state.config.paths.ROOT` anyway, so `.resolve` _should_ always work 🗡

**Comment by SleeplessByte** [comment]
The tests haven't been running for a while (because travis has been broken for over a month). 

> packages/react-static/src/static/components/__tests__/__snapshots__/BodyWithMeta.test.js.snap

This one can be solved by running `yarn test -u` which will update the snapshot. The new output is the expected one 👍 

I think `if (assetsPath !== '' && !assetsPath.endsWith('/')) {` might need to be changed to

```javascript
if (assetsPath && !assetsPath.endsWith('/')) {
```

Can you see if that works? I think you're correct that this isn't YOUR doing 😓 😅

**Comment by jviotti** [comment]
@SleeplessByte Cool, I'll be adding more commits on top of this PR to try to fix the issues. I applied your suggested fix and updated the snapshots.

**Comment by jviotti** [comment]
@SleeplessByte OK, I got it to pass locally \o/ Can you give a look at the 2 commits I added here?

**Comment by jviotti** [comment]
And Travis is green again!

File: packages/react-static/src/static/getRoutes.js (hunk 1 above)
Comment (jviotti) [reply packages/react-static/src/static/getRoutes.js]: @SleeplessByte 

> I think we used to be able to change the ROOT dir

We still can do so. This part of the code only kicks in when the user didn't define a 404 page, and therefore we use the default one that comes with the module. The ROOT dir shouldn't make a difference here.

File: packages/react-static/src/static/__mocks__/defaultConfigProduction.mock.js [hunk 2]
```diff
@@ -19,6 +18,11 @@ export default {
     STATIC_DATA: './root/dist/staticData',
   },
   terser: {
-    terserOptions: {},
+    terserOptions: {
+      compress: {},
```
Comment (SleeplessByte) [inline packages/react-static/src/static/__mocks__/defaultConfigProduction.mock.js]: Can you explain what this fixed/changed? I just don't know 😅

File: packages/react-static/src/static/__tests__/getConfig.test.js [hunk 3]
```diff
@@ -101,7 +101,7 @@ describe('buildConfig', () => {
     test('when no user-supplied assetsPath exists', () => {
       const { config } = buildConfig({}, { assetsPath: undefined })
 
-      expect(config.assetsPath).toBe('')
+      expect(config.assetsPath).toBeUndefined()
```
Comment (SleeplessByte) [inline packages/react-static/src/static/__tests__/getConfig.test.js]: Does this mean that we _no longer_ default to `''`? I like this/your code better btw!

File: packages/react-static/src/static/__mocks__/defaultConfigProduction.mock.js (hunk 2 above)
Comment (jviotti) [reply packages/react-static/src/static/__mocks__/defaultConfigProduction.mock.js]: I hoped you would know :P From what I can tell, the default config includes these nested empty objects, while the obsolete tests were still expecting just `terserOptions: {}`

File: packages/react-static/src/static/__tests__/getConfig.test.js (hunk 3 above)
Comment (jviotti) [reply packages/react-static/src/static/__tests__/getConfig.test.js]: That seems to be what's happening at the moment, just that the tests were outdated. Let me give it a shot first anyways, just in case I can also fix it to be `''`

Comment (jviotti) [reply packages/react-static/src/static/__tests__/getConfig.test.js]: @SleeplessByte Done! It should be correctly fixed now

File: packages/react-static/src/static/__mocks__/defaultConfigProduction.mock.js (hunk 2 above)
Comment (jviotti) [reply packages/react-static/src/static/__mocks__/defaultConfigProduction.mock.js]: Hm, I think I also solved this with my other fix. Let me check

Comment (jviotti) [reply packages/react-static/src/static/__mocks__/defaultConfigProduction.mock.js]: Indeed! Let me cleanup my commits a bit and we should be good to go.

**Comment by jviotti** [comment]
@SleeplessByte Can you take another look now? The diff is much cleaner now :)

**Comment by SleeplessByte** [comment]
Great!! We'll edit those obsolete tests in a future iteration :)
